File: src/services/historyMeta.js
// 이력 메타 파생 — 순수 모듈(DB·HTTP·Date·랜덤 비의존, embargoPolicy.js와 동형 관례).
// ArticleHistory에는 기록 시점에 파생 저장한 표시 제목 컬럼(snapshotTitle — phase 58)이 있고,
// 버전 저장 컬럼은 없으며 상태는 전이 행에만 있다. 이 모듈은 이력 로그 + 스냅샷 항목
// ({ id, snapshotTitle?, markupVersion? })에서 표시용 title/version/status를 파생한다 —
// 저장된 제목이 1차 출처이고, 값이 없는(null) 레거시 항목만 본문(markupVersion)에서 파생한다.
// 조회·필터 결선은 서비스(articleService.queryHistory) 책임 — 필요한 본문은 전부 인자로 주입받는다.

export const MAX_HISTORY_TITLE_LEN = 200;

// markupVersion 스냅샷 → 그 본문의 제목(첫 텍스트 줄, trim, 상한 절단). 파싱 불가/빈 값 → ''.
// 규칙 드리프트(의도된 대가): 아래 파생 규칙을 바꿔도 이미 저장된 행(snapshotTitle 컬럼)은 옛 규칙의
// 값을 유지한다(빈 컬럼만 채우는 부트 백필은 있고 그 백필도 이 함수를 파생 출처로 쓴다 — 단, 이미
// 저장된 값은 재파생·덮어쓰지 않는다) — 규칙을 바꾸면 같은 목록 안에서 행마다 다른 규칙의 제목이 보일 수 있다.
// 규칙은 프론트 bodyTitle(blocksToText(deserialize(body)) 첫 줄)과 동형이어야 한다 —
// 텍스트 블록(type:'text')만 세고, 임베드의 title 같은 필드는 제목으로 오인하지 않는다.
// 깨진 JSON·평문 레거시는 문자열 그대로 취급한다(hasEndMarker의 방어 패턴과 동일).
// 동형의 한계(의도): 200자 상한은 이 파생에만 있고(Article.title은 무상한 — 표시 절단일 뿐),
// text 비문자열·blocks 없는 JSON 객체 같은 serialize() 불능 입력의 처리도 프론트와 다를 수 있다.
export function snapshotTitle(markupVersion) {
  if (markupVersion === null || markupVersion === undefined || markupVersion === '') return '';
  const raw = String(markupVersion);
  let text = raw;
  try {
    const doc = JSON.parse(raw);
    const blocks = Array.isArray(doc) ? doc : (doc && Array.isArray(doc.blocks) ? doc.blocks : null);
    if (blocks) {
      text = blocks
        .filter((b) => b && b.type === 'text' && typeof b.text === 'string')
        .map((b) => b.text)
        .join('\n');
    }
  } catch {
    // 평문 레거시 — text는 이미 raw.
  }
  const firstLine = (text.split('\n')[0] ?? '').trim();
  return firstLine.slice(0, MAX_HISTORY_TITLE_LEN);
}

// 모델(queryByArticle)의 hasSnapshot(숫자 1/0)과 같은 판정 — 버전 증가는 이 플래그만 본다.
function hasSnap(rowItem) {
  return !!rowItem.hasSnapshot;
}

// 전이 행 판정 — status 전이 행만 toStatus를 갖는다(distribute·edit 행은 null). 전용 분기 없음.
function isTransition(rowItem) {
  return rowItem.toStatus !== null && rowItem.toStatus !== undefined && rowItem.toStatus !== '';
}

// 제목 결정 규칙(단일 지점): 저장된 파생 제목이 문자열이면 그대로 쓰고(재파생 금지 — 평문 제목에
// snapshotTitle()을 다시 적용하면 '[1,2]'처럼 JSON으로 파싱되는 제목이 파괴된다), 그렇지 않을 때만
// markupVersion에서 파생한다. ''(빈 문자열)도 유효한 저장값이다 — 폴백하지 않는다.
function resolveSnapshotTitle(item) {
  if (!item) return '';
  if (typeof item.snapshotTitle === 'string') return item.snapshotTitle;
  return snapshotTitle(item.markupVersion);
}

// rows: articleHistoryModel.queryByArticle() 결과 — id DESC(최신순) 전제. 순서 판단은 배열 위치가
//   아니라 id로 한다(같은 createdAt이 여러 건일 수 있다). 반환 순서는 입력 순서 그대로(재정렬 금지).
// snapshots: [{ id, snapshotTitle?, markupVersion? }] 배열 하나(undefined/빈 배열 허용).
//   신규 항목은 snapshotTitle(저장된 파생 제목) + markupVersion: null, 레거시 항목은
//   snapshotTitle: null(또는 키 없음) + markupVersion: <본문> — 같은 배열에 혼재할 수 있다.
// options.v1Body: 스냅샷이 '한 건도 없을 때만' 쓰는 v1 본문(= 현재 Article.markupVersion).
//   스냅샷 0건 = 최초 저장 이후 본문이 안 바뀌었다는 뜻이므로 현재 본문 = v1 본문(동치).
//   스냅샷이 1건이라도 있으면 현재 본문은 최신 버전이지 v1이 아니므로 무시한다.
// 반환: 입력과 같은 길이·같은 순서의 새 배열. 각 원소 = 입력 행 얕은 복사 + { title, version, status }.
//   version은 1-base 정수(최초 저장 본문 = v1, 스냅샷 행마다 증가 — 스냅샷 행 자신은 증가 후 값).
//   title·status는 값이 없으면 ''(표시 폴백 '—'는 뷰 책임).
//   markupVersion·snapshotTitle은 반환 행에 싣지 않는다(응답 shape 불변 — 방어적 제거).
export function decorateHistoryRows(rows, snapshots, { v1Body } = {}) {
  const input = Array.isArray(rows) ? rows : [];
  const snapMap = new Map(
    (Array.isArray(snapshots) ? snapshots : []).map((s) => [s.id, s]),
  );

  // 계산은 오래된 순(id ASC) — 입력 배열은 건드리지 않는다.
  const asc = [...input].sort((a, b) => a.id - b.id);
  const anySnapshot = input.some(hasSnap);
  // 스냅샷 0건일 때만 v1Body로 v1 구간 제목을 파생한다(그 외에는 무시 — 위 주석 참조).
  const v1Title = anySnapshot ? '' : snapshotTitle(v1Body);

  // 역승계용: 가장 오래된 전이의 fromStatus — 그보다 앞선 행의 상태는 그 전이 행이 알고 있다.
  const firstTransition = asc.find(isTransition);
  const backfillStatus = firstTransition ? (firstTransition.fromStatus ?? '') : '';

  let version = 1;
  let title = v1Title;
  let status = null; // 아직 전이를 못 만난 구간 — 마지막에 backfillStatus로 채운다.
  const derivedById = new Map();
  for (const r of asc) {
    if (hasSnap(r)) {
      version += 1; // 스냅샷 행 자신은 '증가 후' 값(그 편집이 새 버전을 만든다).
      title = resolveSnapshotTitle(snapMap.get(r.id));
    }
    if (isTransition(r)) status = r.toStatus;
    derivedById.set(r.id, { version, title, status: status ?? backfillStatus });
  }

  return input.map((r) => {
    const out = { ...r, ...derivedById.get(r.id) };
    delete out.markupVersion; // 목록은 blob 없는 경량 계약 — 어떤 입력에서도 싣지 않는다.
    delete out.snapshotTitle; // 표시 값의 단일 출처는 title — 저장 컬럼 값을 응답에 싣지 않는다(방어).
    return out;
  });
}

File: server-spring/src/main/java/harness/news/service/HistoryMeta.java
package harness.news.service;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 이력 표시 값의 파생(title·version·status) — 리포 루트 {@code src/services/historyMeta.js}의 이식
 * (순수 모듈: DB·HTTP·시계·난수 비의존).
 *
 * <p>{@code ArticleHistory}에는 기록 시점에 파생 저장한 표시 제목 컬럼({@code snapshotTitle})이 있고,
 * 버전 컬럼은 없으며 상태는 전이 행에만 있다. 이 모듈은 이력 행 + 스냅샷 항목에서 표시용 값을 만든다 —
 * <b>저장된 제목이 1차 출처</b>이고 값이 없는 레거시 항목만 본문에서 파생한다.
 *
 * <p>제목 파생은 {@code type === 'text'}인 블록만 센다. 마커 판정({@link EndMarker})이 모든 블록의
 * {@code text}를 보는 것과 다르다 — 합치면 임베드의 {@code title} 같은 필드가 제목으로 오인된다.
 */
public final class HistoryMeta {

	/** 표시 제목 상한. {@code Article.title}은 무상한이며 이 절단은 이력 표시에만 있다. */
	public static final int MAX_HISTORY_TITLE_LEN = 200;

	private static final String BLOCKS = "blocks";

	private static final String TEXT = "text";

	private static final String TYPE = "type";

	// --- 제목 파생 -----------------------------------------------------------------------------

	/**
	 * 본문 스냅샷 → 표시 제목(첫 텍스트 줄, trim, 200자 절단). 파싱 불가·빈 값은 {@code ''}이다.
	 *
	 * <p>프론트 {@code bodyTitle}과 동형이어야 한다 — 텍스트 블록만 세고, 깨진 JSON·평문 레거시는
	 * 문자열 그대로 취급한다.
	 *
	 * <p>다듬기는 {@link NodeString#trim(String)}이다({@link String#strip()}·{@link String#trim()}
	 * 금지 — 셋의 공백 집합이 서로 다르다). 이 값은 {@code ArticleHistory.snapshotTitle}로 <b>영속</b>되고
	 * 이력 응답의 {@code title}로 나가므로, 집합이 갈리면 같은 편집에 두 서버가 다른 값을 쓴다.
	 * 붙여넣기 본문의 NBSP는 실제로 흔하고 계약 픽스처는 ASCII 제목만 쓴다.
	 */
	public static String snapshotTitle(Object markupVersion) {
		if (markupVersion == null) {
			return "";
		}
		String raw = String.valueOf(markupVersion);
		if (raw.isEmpty()) {
			return "";
		}

		String text = textOf(raw);
		int lineEnd = text.indexOf('\n');
		String firstLine = NodeString.trim((lineEnd < 0) ? text : text.substring(0, lineEnd));
		return (firstLine.length() > MAX_HISTORY_TITLE_LEN)
				? firstLine.substring(0, MAX_HISTORY_TITLE_LEN) : firstLine;
	}

	/** 블록 문서(최상위 배열 또는 {@code blocks} 배열)면 텍스트 블록만 이어 붙이고, 아니면 원문이다. */
	private static String textOf(String raw) {
		Object doc = MarkupJson.parseOrNull(raw);
		List<?> blocks = null;
		if (doc instanceof List<?> list) {
			blocks = list;
		}
		else if (doc instanceof Map<?, ?> map && map.get(BLOCKS) instanceof List<?> list) {
			blocks = list;
		}
		if (blocks == null) {
			return raw;
		}

		StringBuilder joined = new StringBuilder();
		int kept = 0;
		for (Object candidate : blocks) {
			if (!(candidate instanceof Map<?, ?> block) || !TEXT.equals(block.get(TYPE))
					|| !(block.get(TEXT) instanceof String text)) {
				continue; // 텍스트가 아닌 블록·비문자열 text는 아예 버린다(빈 줄도 남기지 않는다).
			}
			if (kept > 0) {
				joined.append('\n'); // 남은 블록이 빈 문자열이어도 줄은 센다(Node join 동형).
			}
			joined.append(text);
			kept += 1;
		}
		return joined.toString();
	}

	// --- 행 파생 -------------------------------------------------------------------------------

	/**
	 * 이력 행에 표시용 {@code title}·{@code version}·{@code status}를 얹는다.
	 *
	 * <p>규칙(정본과 1:1):
	 * <ol>
	 *   <li>계산은 <b>id 오름차순</b>(같은 {@code createdAt}이 여럿일 수 있어 배열 위치를 믿지 않는다),
	 *       반환은 <b>입력 순서 그대로</b>.</li>
	 *   <li>스냅샷 보유 행에서 version을 <b>먼저 올리고 그 값을 그 행에</b> 준다(그 편집이 새 버전을 만든다).</li>
	 *   <li>전이 행({@code toStatus}가 비어 있지 않은 행)에서 상태를 갱신한다.</li>
	 *   <li>첫 전이 이전 구간은 그 전이의 {@code fromStatus}로 <b>역승계</b>한다.</li>
	 *   <li>저장된 파생 제목이 문자열이면 재파생하지 않는다({@code ''}도 유효한 저장값이다).</li>
	 *   <li>{@code v1Body}는 스냅샷이 <b>한 건도 없을 때만</b> 쓴다(그 외에는 현재 본문이 v1이 아니다).</li>
	 *   <li>반환 행에 {@code markupVersion}·{@code snapshotTitle}을 싣지 않는다(응답 shape 불변).</li>
	 * </ol>
	 *
	 * @param rows 이력 행(id DESC 전제이지만 순서에 의존하지 않는다). 입력은 변형하지 않는다.
	 * @param snapshots {@code {id, snapshotTitle?, markupVersion?}} 항목들. 신규·레거시가 섞일 수 있다.
	 * @param v1Body 스냅샷이 없을 때만 쓰는 v1 본문(= 현재 {@code Article.markupVersion})
	 */
	public static List<Map<String, Object>> decorateHistoryRows(List<? extends Map<String, Object>> rows,
			List<? extends Map<String, Object>> snapshots, Object v1Body) {
		List<? extends Map<String, Object>> input = (rows == null) ? List.of() : rows;

		Map<Long, Map<String, Object>> snapshotById = new HashMap<>();
		for (Map<String, Object> snapshot : (snapshots == null) ? List.<Map<String, Object>>of() : snapshots) {
			snapshotById.put(idOf(snapshot.get("id")), snapshot);
		}

		List<Map<String, Object>> ascending = new ArrayList<>(input);
		ascending.sort(Comparator.comparing((row) -> idOf(row.get("id")),
				Comparator.nullsFirst(Comparator.naturalOrder())));

		boolean anySnapshot = input.stream().anyMatch(HistoryMeta::hasSnapshot);
		// 스냅샷이 1건이라도 있으면 현재 본문은 최신 버전이지 v1이 아니다 — 그때는 v1Body를 무시한다.
		String v1Title = anySnapshot ? "" : snapshotTitle(v1Body);
		String backfillStatus = backfillStatus(ascending);

		Map<Long, Derived> derivedById = new HashMap<>();
		int version = 1;
		String title = v1Title;
		String status = null; // 아직 전이를 못 만난 구간 — backfillStatus로 채운다.
		for (Map<String, Object> row : ascending) {
			if (hasSnapshot(row)) {
				version += 1;
				title = resolveTitle(snapshotById.get(idOf(row.get("id"))));
			}
			if (isTransition(row)) {
				status = String.valueOf(row.get("toStatus"));
			}
			derivedById.put(idOf(row.get("id")), new Derived(version, title, (status == null) ? backfillStatus : status));
		}

		List<Map<String, Object>> out = new ArrayList<>(input.size());
		for (Map<String, Object> row : input) {
			Map<String, Object> decorated = new LinkedHashMap<>(row);
			Derived derived = derivedById.get(idOf(row.get("id")));
			decorated.put("version", derived.version());
			decorated.put("title", derived.title());
			decorated.put("status", derived.status());
			decorated.remove("markupVersion"); // 목록은 blob 없는 경량 계약이다.
			decorated.remove("snapshotTitle"); // 표시 값의 단일 출처는 title이다.
			out.add(decorated);
		}
		return out;
	}

	private record Derived(int version, String title, String status) {
	}

	/** 가장 오래된 전이의 {@code fromStatus} — 그보다 앞선 행의 상태는 그 전이 행이 알고 있다. */
	private static String backfillStatus(List<Map<String, Object>> ascending) {
		for (Map<String, Object> row : ascending) {
			if (isTransition(row)) {
				Object from = row.get("fromStatus");
				return (from == null) ? "" : String.valueOf(from);
			}
		}
		return "";
	}

	/** 저장된 파생 제목이 문자열이면 그대로 쓴다 — 재파생하면 평문 제목이 파괴된다. */
	private static String resolveTitle(Map<String, Object> snapshot) {
		if (snapshot == null) {
			return "";
		}
		if (snapshot.get("snapshotTitle") instanceof String stored) {
			return stored;
		}
		return snapshotTitle(snapshot.get("markupVersion"));
	}

	/** 모델이 주는 {@code hasSnapshot}(정수 1/0)의 참 판정 — 버전 증가는 이 플래그만 본다. */
	private static boolean hasSnapshot(Map<String, Object> row) {
		Object flag = row.get("hasSnapshot");
		if (flag instanceof Number number) {
			return number.longValue() != 0L;
		}
		if (flag instanceof Boolean bool) {
			return bool;
		}
		return (flag instanceof String text) && !text.isEmpty();
	}

	/** 전이 행 판정 — status 전이 행만 {@code toStatus}를 갖는다(edit·distribute 행은 비어 있다). */
	private static boolean isTransition(Map<String, Object> row) {
		Object to = row.get("toStatus");
		return to != null && !String.valueOf(to).isEmpty();
	}

	/** 순서·대응의 기준은 배열 위치가 아니라 id다. 정수가 아닌 id는 값이 없는 것으로 본다. */
	private static Long idOf(Object id) {
		if (id instanceof Number number) {
			return number.longValue();
		}
		if (id instanceof String text) {
			try {
				return Long.valueOf(text);
			}
			catch (NumberFormatException ex) {
				return null;
			}
		}
		return null;
	}

	private HistoryMeta() {
	}
}
